File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/instructions/SetKaigiRelayManifestInstruction.java
package org.hyperledger.iroha.android.model.instructions;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/** Typed builder for {@code SetKaigiRelayManifest} instructions. */
public final class SetKaigiRelayManifestInstruction implements InstructionTemplate {

  private static final String ACTION = "SetKaigiRelayManifest";
  private static final java.util.Set<String> ALLOWED_ARGUMENTS =
      KaigiInstructionUtils.argumentSet(
          "action",
          "call.domain_id",
          "call.call_name",
          "relay_manifest.expiry_ms");

  private final KaigiInstructionUtils.CallId callId;
  private final KaigiInstructionUtils.RelayManifest relayManifest;
  private final Map<String, String> arguments;

  private SetKaigiRelayManifestInstruction(final Builder builder) {
    this(builder, builder.canonicalArguments());
  }

  private SetKaigiRelayManifestInstruction(
      final Builder builder, final Map<String, String> argumentOrder) {
    this.callId = builder.callId;
    this.relayManifest = builder.buildRelayManifest();
    this.arguments = Collections.unmodifiableMap(new LinkedHashMap<>(argumentOrder));
  }

  public KaigiInstructionUtils.CallId callId() {
    return callId;
  }

  public KaigiInstructionUtils.RelayManifest relayManifest() {
    return relayManifest;
  }

  @Override
  public InstructionKind kind() {
    return InstructionKind.CUSTOM;
  }

  @Override
  public Map<String, String> toArguments() {
    return arguments;
  }

  public static SetKaigiRelayManifestInstruction fromArguments(final Map<String, String> arguments) {
    KaigiInstructionUtils.requireKnownArguments(
        arguments, ALLOWED_ARGUMENTS, "relay_manifest.hop.");
    KaigiInstructionUtils.requireAction(arguments, ACTION);
    final Builder builder = builder();
    builder.setCallId(KaigiInstructionUtils.parseCallId(arguments, "call"));
    builder.setRelayManifest(KaigiInstructionUtils.parseRelayManifest(arguments, "relay_manifest"));
    return builder.build();
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof SetKaigiRelayManifestInstruction other)) {
      return false;
    }
    return Objects.equals(callId.domainId(), other.callId.domainId())
        && Objects.equals(callId.callName(), other.callId.callName())
        && CreateKaigiInstruction.relayManifestEquals(relayManifest, other.relayManifest);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        callId.domainId(),
        callId.callName(),
        CreateKaigiInstruction.relayManifestHash(relayManifest));
  }

  public static final class Builder {
    private KaigiInstructionUtils.CallId callId;
    private Long relayManifestExpiry;
    private final List<KaigiInstructionUtils.RelayManifestHop> relayManifestHops = new ArrayList<>();

    private Builder() {}

    public Builder setCallId(final String domainId, final String callName) {
      this.callId = new KaigiInstructionUtils.CallId(domainId, callName);
      return this;
    }

    public Builder setCallId(final KaigiInstructionUtils.CallId callId) {
      this.callId = Objects.requireNonNull(callId, "callId");
      return this;
    }

    public Builder clearRelayManifest() {
      relayManifestExpiry = null;
      relayManifestHops.clear();
      return this;
    }

    public Builder setRelayManifestExpiryMs(final Long expiryMs) {
      this.relayManifestExpiry = expiryMs;
      return this;
    }

    public Builder addRelayManifestHop(
        final String relayId, final String hpkePublicKeyBase64, final int weight) {
      if (relayId == null || relayId.isBlank()) {
        throw new IllegalArgumentException("relayId must not be blank");
      }
      if (relayManifestHops.size() >= KaigiInstructionUtils.KAIGI_RELAY_MANIFEST_MAX_HOPS_V1) {
        throw new IllegalArgumentException(
            "relay manifest must not contain more than "
                + KaigiInstructionUtils.KAIGI_RELAY_MANIFEST_MAX_HOPS_V1
                + " hops");
      }
      final String normalizedKey =
          KaigiInstructionUtils.requireHpkePublicKeyBase64(
              hpkePublicKeyBase64, "hpkePublicKey");
      if (weight < 1 || weight > 0xFF) {
        throw new IllegalArgumentException("relay hop weight must be between 1 and 255");
      }
      relayManifestHops.add(
          new KaigiInstructionUtils.RelayManifestHop()
              .withRelayId(relayId)
              .withHpkePublicKey(normalizedKey)
              .withWeight(weight));
      return this;
    }

    public Builder addRelayManifestHop(
        final String relayId, final byte[] hpkePublicKey, final int weight) {
      return addRelayManifestHop(
          relayId,
          KaigiInstructionUtils.toHpkePublicKeyBase64(hpkePublicKey, "hpkePublicKey"),
          weight);
    }

    public Builder setRelayManifest(final KaigiInstructionUtils.RelayManifest manifest) {
      relayManifestExpiry = null;
      relayManifestHops.clear();
      if (manifest != null) {
        relayManifestExpiry = manifest.expiryMs();
        for (KaigiInstructionUtils.RelayManifestHop hop : manifest.hops()) {
          relayManifestHops.add(
              new KaigiInstructionUtils.RelayManifestHop(
                  hop.relayId(), hop.hpkePublicKey(), hop.weight()));
        }
      }
      return this;
    }

    public SetKaigiRelayManifestInstruction build() {
      if (callId == null) {
        throw new IllegalStateException("callId must be provided");
      }
      return new SetKaigiRelayManifestInstruction(this);
    }

    private KaigiInstructionUtils.RelayManifest buildRelayManifest() {
      if (relayManifestHops.isEmpty() && relayManifestExpiry == null) {
        return null;
      }
      final List<KaigiInstructionUtils.RelayManifestHop> copy =
          new ArrayList<>(relayManifestHops.size());
      for (KaigiInstructionUtils.RelayManifestHop hop : relayManifestHops) {
        copy.add(
            new KaigiInstructionUtils.RelayManifestHop(
                hop.relayId(), hop.hpkePublicKey(), hop.weight()));
      }
      return KaigiInstructionUtils.validateRelayManifest(
          new KaigiInstructionUtils.RelayManifest(
              relayManifestExpiry, Collections.unmodifiableList(copy)));
    }

    private Map<String, String> canonicalArguments() {
      final Map<String, String> args = new LinkedHashMap<>();
      args.put("action", ACTION);
      KaigiInstructionUtils.appendCallId(callId, args, "call");
      KaigiInstructionUtils.appendRelayManifest(buildRelayManifest(), args, "relay_manifest");
      return args;
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/instructions/SetKaigiRelayManifestInstruction.kt
package org.hyperledger.iroha.sdk.core.model.instructions

/** Typed builder for `SetKaigiRelayManifest` instructions. */
class SetKaigiRelayManifestInstruction internal constructor(
    @JvmField val callId: KaigiInstructionUtils.CallId,
    @JvmField val relayManifest: KaigiInstructionUtils.RelayManifest?,
    arguments: Map<String, String>,
) : InstructionTemplate {

    override val arguments: Map<String, String> =
        KaigiInstructionUtils.immutableArguments(arguments)

    override val kind: InstructionKind get() = InstructionKind.CUSTOM

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SetKaigiRelayManifestInstruction) return false
        return callId.domainId == other.callId.domainId
            && callId.callName == other.callId.callName
            && CreateKaigiInstruction.relayManifestEquals(relayManifest, other.relayManifest)
    }

    override fun hashCode(): Int {
        var result = callId.domainId.hashCode()
        result = 31 * result + callId.callName.hashCode()
        result = 31 * result + CreateKaigiInstruction.relayManifestHash(relayManifest)
        return result
    }

    class Builder internal constructor() {
        private var callId: KaigiInstructionUtils.CallId? = null
        private var relayManifestExpiry: Long? = null
        private val relayManifestHops: MutableList<KaigiInstructionUtils.RelayManifestHop> = mutableListOf()

        fun setCallId(domainId: String, callName: String) = apply {
            this.callId = KaigiInstructionUtils.CallId(domainId, callName)
        }

        fun setCallId(callId: KaigiInstructionUtils.CallId) = apply {
            this.callId = requireNotNull(callId) { "callId" }
        }

        fun clearRelayManifest() = apply {
            relayManifestExpiry = null
            relayManifestHops.clear()
        }

        fun setRelayManifestExpiryMs(expiryMs: Long?) = apply {
            this.relayManifestExpiry = expiryMs
        }

        fun addRelayManifestHop(relayId: String, hpkePublicKeyBase64: String, weight: Int) = apply {
            require(relayId.isNotBlank()) { "relayId must not be blank" }
            require(relayManifestHops.size < KaigiInstructionUtils.KAIGI_RELAY_MANIFEST_MAX_HOPS_V1) {
                "relay manifest must not contain more than " +
                    "${KaigiInstructionUtils.KAIGI_RELAY_MANIFEST_MAX_HOPS_V1} hops"
            }
            val normalizedKey = KaigiInstructionUtils.requireHpkePublicKeyBase64(
                hpkePublicKeyBase64,
                "hpkePublicKey",
            )
            require(weight in 1..0xFF) { "relay hop weight must be between 1 and 255" }
            relayManifestHops.add(
                KaigiInstructionUtils.RelayManifestHop(relayId, normalizedKey, weight),
            )
        }

        fun addRelayManifestHop(relayId: String, hpkePublicKey: ByteArray, weight: Int) = apply {
            addRelayManifestHop(
                relayId,
                KaigiInstructionUtils.toHpkePublicKeyBase64(hpkePublicKey, "hpkePublicKey"),
                weight,
            )
        }

        fun setRelayManifest(manifest: KaigiInstructionUtils.RelayManifest?) = apply {
            relayManifestExpiry = null
            relayManifestHops.clear()
            if (manifest != null) {
                relayManifestExpiry = manifest.expiryMs
                for (hop in manifest.hops) {
                    relayManifestHops.add(
                        KaigiInstructionUtils.RelayManifestHop(hop.relayId, hop.hpkePublicKey, hop.weight),
                    )
                }
            }
        }

        fun build(): SetKaigiRelayManifestInstruction {
            val id = checkNotNull(callId) { "callId must be provided" }
            val manifest = buildRelayManifest()
            return SetKaigiRelayManifestInstruction(id, manifest, canonicalArguments(id, manifest))
        }

        private fun buildRelayManifest(): KaigiInstructionUtils.RelayManifest? {
            if (relayManifestHops.isEmpty() && relayManifestExpiry == null) return null
            val copy = relayManifestHops.map {
                KaigiInstructionUtils.RelayManifestHop(it.relayId, it.hpkePublicKey, it.weight)
            }
            return KaigiInstructionUtils.validateRelayManifest(
                KaigiInstructionUtils.RelayManifest(relayManifestExpiry, copy.toList()),
            )
        }

        private fun canonicalArguments(
            id: KaigiInstructionUtils.CallId,
            manifest: KaigiInstructionUtils.RelayManifest?,
        ): Map<String, String> {
            val args = LinkedHashMap<String, String>()
            args["action"] = ACTION
            KaigiInstructionUtils.appendCallId(id, args, "call")
            KaigiInstructionUtils.appendRelayManifest(manifest, args, "relay_manifest")
            return args
        }
    }

    companion object {
        private const val ACTION = "SetKaigiRelayManifest"

        @JvmStatic
        fun builder(): Builder = Builder()

        @JvmStatic
        fun fromArguments(arguments: Map<String, String>): SetKaigiRelayManifestInstruction {
            KaigiInstructionUtils.requireKnownArguments(
                arguments,
                SET_KAIGI_RELAY_MANIFEST_ARGUMENTS,
                "relay_manifest.hop.",
            )
            KaigiInstructionUtils.requireAction(arguments, ACTION)
            val callId = KaigiInstructionUtils.parseCallId(arguments, "call")
            val relayManifest = KaigiInstructionUtils.parseRelayManifest(arguments, "relay_manifest")
            return builder()
                .setCallId(callId)
                .setRelayManifest(relayManifest)
                .build()
        }

        private val SET_KAIGI_RELAY_MANIFEST_ARGUMENTS = setOf(
            "action",
            "call.domain_id",
            "call.call_name",
            "relay_manifest.expiry_ms",
        )
    }
}
